pkg/operator/target_status.go

```diff
@@ -0,0 +1,973 @@
+	m.tryMutex.Unlock()
+}
+
+// Tries to lock the mutex. Returns true if the mutex was locked. The mutex could be unlocked but still return false.
+func (m *tryMutex) TryLock() bool {
+	var success bool
+	m.tryMutex.Lock()
+	if !m.locked {
+		success = true
+		m.locked = true
+	} else {
+		success = false
+	}
+	m.tryMutex.Unlock()
+
+	if success {
+		m.mutex.Lock()
+	}
+	return success
+}
+
+const (
+	// How often to do a full refresh of cached Kubernetes data.
+	informerResyncInterval = time.Minute * 15
+	// How many targets to keep in each group.
+	maxSampleTargetSize = 5
+)
+
+var (
+	targetStatusDuration = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "prometheus_engine_target_status_duration",
+		Help: "A metric indicating how long it took to fetch the complete target status.",
+	}, []string{})
+)
+
+// Wraps a time.Ticker so we can replace for unit testing.
+type Ticker interface {
+	// Stops the ticker from ticking.
+	Stop()
+	// Returns the channel receiving ticks.
+	Channel() <-chan time.Time
+}
+
+type timeTicker time.Ticker
+
+func NewTicker(d time.Duration) Ticker {
+	data := timeTicker(*time.NewTicker(d))
+	return &data
+}
+
+func (t *timeTicker) Stop() {
+	ticker := time.Ticker(*t)
+	ticker.Stop()
+}
+
+func (t *timeTicker) Channel() <-chan time.Time {
+	ticker := time.Ticker(*t)
+	return ticker.C
+}
+
+type TargetStatusProcessor interface {
+	Start(ctx context.Context, tickerBuilder func() Ticker) error
+}
+
+type dataManagerBuilder interface {
+	build(stopCh chan struct{}) (dataManager, error)
+}
+
+// Manages fetching and patching Kubernetes data.
+type dataManager interface {
+	// Returns all pods running Prometheus, mapped with port as the key.
+	getPrometheusPods() (map[int32][]*corev1.Pod, error)
+
+	getPodMonitorings() ([]*monitoringv1.PodMonitoring, error)
+	getClusterPodMonitorings() ([]*monitoringv1.ClusterPodMonitoring, error)
+
+	patchPodMonitoring(ctx context.Context, podMonitoring *monitoringv1.PodMonitoring) error
+	patchClusterPodMonitoring(ctx context.Context, podMonitoring *monitoringv1.ClusterPodMonitoring) error
+
+	// Returns the target fetcher, with an zero-index number `thread` that signifies which goroutine this is for.
+	getTargetFetcher(ctx context.Context, thread uint16) (targetFetcher, error)
+}
+
+// Responsible for fetching the targets given a pod. One is created per goroutine.
+type targetFetcher interface {
+	getTarget(ctx context.Context, logger logr.Logger, port int32, pod *corev1.Pod) (*prometheusv1.TargetsResult, error)
+}
+
+type targetStatusProcessor struct {
+	logger             logr.Logger
+	opts               Options
+	dataManagerBuilder dataManagerBuilder
+}
+
+func NewTargetStatusPoller(logger logr.Logger, restConfig *rest.Config, registry prometheus.Registerer, opts Options) (TargetStatusProcessor, error) {
+	builder, err := newInformerDataManagerBuilder(restConfig, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return newTargetStatusPoller(logger, opts, registry, builder)
```

Comment:
any reason for this nested private function?

Reply:
The private version is used internally for unit tests.